server/middleware/rest.js

```diff
@@ -57,9 +59,16 @@ function rest() {
         preHandlers.push(loopback.token({ model: AccessToken }));
       }
     }
-
-    async.eachSeries(preHandlers.concat(restHandler), function(handler, done) {
-      handler(req, res, done);
-    }, next);
+    if (preHandlers.length > 0) {
+      // Build a router to chain all handlers
+      var router = loopback.Router();
+      for (var i = 0, n = preHandlers.length; i < n; i++) {
+        router.use(preHandlers[i]);
+      }
+      router.use(restHandler);
+      return router(req, res, next);
```

Comment:
I am surprised that building a new Router for each request is faster than running `async.each`. Did the benchmark confirm your assumption?

FWIW, since the list of pre-handlers is static, it should be possible to cache the Router instance and just wrap `restHandler` in a function, so that it's dynamically fetched for each request.

```js
var router = loopback.Router();
for (var i = 0, n = preHandlers.length; i < n; i++) {
  router.use(preHandlers[i]);
}
router.use(function(req, res, next) { return app.handler('rest')(res, req, next); });
```

Of course, you need to run the benchmark to see which version is faster.